# Inventory Management System

## Project Overview
This project will help you build a complete inventory management system that can read product data from a CSV file, track inventory levels, calculate values, and generate reports.  You are to fill in the cells with the required code based upon the TODO instructions.  The code in steps 13 and 14 has been provided for you to get you started.



## Step 1: Import Required Modules

We'll use the `csv` module for reading CSV files properly.


In [20]:
# Import the csv module for reading CSV files and print a message confirming the import was successful.
import csv
print("CSV module successfully imported.")

# Path to the inventory CSV file. Define once here so it can be referenced by name throughout the notebook.
CSV_FILENAME = 'Data/inventory_data.csv'

CSV module successfully imported.


## Preview Data

Before we load the data into our dictionary, let's preview the first few rows of the CSV file. This helps us see the structure of the file and confirm it has a header row that needs to be skipped.


In [21]:
# Preview the first 5 rows of the CSV file as a table
with open(CSV_FILENAME, 'r') as f:
    reader = csv.reader(f)
    for i, row in enumerate(reader):
        if i == 0:
            # Print header with specific formatting
            print(f"{row[0]:<30} {row[1]:<15} {row[2]}")
            print("-" * 55)
        elif i < 5:
            # Print rows with matching formatting
            print(f"{row[0]:<30} {row[1]:<15} {row[2]}")
        else:
            break
# print the total number of rows in the file (excluding header) and print the data types of each column to confirm they are all strings at this point
with open(CSV_FILENAME, 'r') as f:
    reader = csv.reader(f)
    next(reader)  # skip header
    rows = [r for r in reader if r]  # skip any blank lines
    print(f"\nTotal number of products (rows): {len(rows)}")
    print(f"Data types of columns: Product Name: {type(rows[0][0])}, Quantity: {type(rows[0][1])}, Price: {type(rows[0][2])}")

# if the data types have all numbers as strings, print a message confirming that we will need to convert quantity to int and price to float when loading the data into the dictionary
if all(isinstance(row[1], str) and isinstance(row[2], str) for row in rows):
    print("\nAll quantity and price values are currently strings. We will need to convert quantity to int and price to float when loading the data into the dictionary.")

Product                        Quantity        Price
-------------------------------------------------------
Laptop                         15              899.99
Mouse                          45              24.99
Keyboard                       32              49.99
Monitor                        8               299.99

Total number of products (rows): 15
Data types of columns: Product Name: <class 'str'>, Quantity: <class 'str'>, Price: <class 'str'>

All quantity and price values are currently strings. We will need to convert quantity to int and price to float when loading the data into the dictionary.


## Step 2: Load Inventory Data from CSV

Create a function to read the inventory data from the CSV file and store it in a dictionary. The dictionary should use product names as keys and store quantity and price as values.


In [22]:
# ---------------------------------------------------------------------------
# Function: load_inventory(filename)
#
# Purpose:
#   Reads inventory data from the given CSV file and returns a nested
#   dictionary where each key is a product name and each value is another
#   dictionary with 'quantity' (int) and 'price' (float).
#
# Parameters:
#   filename (str): path to the CSV file (e.g., the CSV_FILENAME variable
#                   defined in Step 1).
#
# Returns:
#   dict: inventory data in the form
#         {product_name: {'quantity': int, 'price': float}, ...}
#         Returns an empty dict if the file is not found.
#
# Notes:
#   - The first row of the CSV is treated as a header and skipped.
#   - Quantity is converted from string to int; price is converted to float.
#   - Wrapped in try/except to gracefully handle FileNotFoundError.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called in the next cell.
# ---------------------------------------------------------------------------
def load_inventory(filename):
    inventory = {}

    try:
        # open the file in read mode
        f = open(filename, 'r')
        reader = csv.reader(f)       # create a csv reader object
        next(reader)                 # skip the header row
        print(f"File '{filename}' opened successfully.")

        # Print a table header so each product's processing info shows up as
        # a row with columns: Product Name | Processing | Quantity | Price | Status
        print(f"{'Product Name':<25} {'Processing':<25} {'Quantity':<10} {'Price':<10} {'Status'}")
        print("-" * 95)

        for row in reader:
            if not row:                  # skip any blank lines in the CSV
                continue
            product_name = row[0]        # first column is product name
            quantity = int(row[1])       # second column is quantity (convert to int)
            price = float(row[2])        # third column is price (convert to float)
            # store in nested dictionary format
            inventory[product_name] = {'quantity': quantity, 'price': price}
            # Print one table row per product instead of separate lines
            print(f"{product_name:<25} {'Processing product':<25} {quantity:<10} {'$' + format(price, '.2f'):<10} {'Added to inventory'}")

        f.close()  # close the file when done
        print("-" * 95)
        print(f"File '{filename}' closed successfully.")
        print(f"Inventory loaded successfully.")

    except FileNotFoundError:
        # if file doesn't exist, print error and return empty dictionary
        print(f"Error: File '{filename}' not found.")

    return inventory

# Confirmation that the cell ran and the function is now available to call.
print("load_inventory function defined and ready to use.")

load_inventory function defined and ready to use.


### Load and Preview Dictionary
After loading the data, the inventory dictionary will use product names as keys. Here is a preview of the keys:


In [23]:
# Example of loading the inventory and previewing the keys
inventory = load_inventory(CSV_FILENAME)

if inventory:
    print(f"Product names used as keys: {list(inventory.keys())[:3]}...")
    print(f"Total unique products loaded: {len(inventory)}")


File 'Data/inventory_data.csv' opened successfully.
Product Name              Processing                Quantity   Price      Status
-----------------------------------------------------------------------------------------------
Laptop                    Processing product        15         $899.99    Added to inventory
Mouse                     Processing product        45         $24.99     Added to inventory
Keyboard                  Processing product        32         $49.99     Added to inventory
Monitor                   Processing product        8          $299.99    Added to inventory
Headphones                Processing product        12         $79.99     Added to inventory
Webcam                    Processing product        5          $89.99     Added to inventory
USB Drive                 Processing product        50         $12.99     Added to inventory
External Hard Drive       Processing product        18         $129.99    Added to inventory
Tablet                    P

### Print Inventory as a Table and Confirm Column Data Types
After loading, print the inventory as a formatted table and confirm the data type stored in each column.


In [24]:
# Print the loaded inventory as a table (same format as the Preview Data cell),
# then confirm the data type of each column.
if inventory:
    # Header row - matches the Preview Data cell's column widths (<30, <15)
    print(f"{'Product Name':<30} {'Quantity':<15} {'Price'}")
    print("-" * 55)
    # Data rows - one product per line, formatted like the preview table
    # Sort product names alphabetically (case-insensitive) for display
    for product_name in sorted(inventory.keys(), key=str.lower):
        details = inventory[product_name]
        print(f"{product_name:<30} {details['quantity']:<15} ${details['price']:.2f}")

    # Confirm data types per column using the first item as a sample
    first_product = next(iter(inventory))
    first_details = inventory[first_product]
    print(f"\nData types of columns: Product Name: {type(first_product)}, Quantity: {type(first_details['quantity'])}, Price: {type(first_details['price'])}")


Product Name                   Quantity        Price
-------------------------------------------------------
Cable                          28              $9.99
Camera                         3               $1299.99
Charger                        35              $19.99
External Hard Drive            18              $129.99
Headphones                     12              $79.99
Keyboard                       32              $49.99
Laptop                         15              $899.99
Microphone                     7               $199.99
Monitor                        8               $299.99
Mouse                          45              $24.99
Smartphone                     6               $699.99
Speaker                        14              $149.99
Tablet                         22              $449.99
USB Drive                      50              $12.99
Webcam                         5               $89.99

Data types of columns: Product Name: <class 'str'>, Quantity: <class 'in

## Step 3: Display All Inventory Items

Create a function to display all items in the inventory in a formatted table.


In [25]:
# ---------------------------------------------------------------------------
# Function: display_inventory(inventory)
#
# Purpose:
#   Prints all items in the inventory as a formatted table with three
#   columns: Product, Quantity, and Price.
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#
# Returns:
#   None. Output is printed directly to the console.
#
# Notes:
#   - Column widths match the Preview Data cell (<30, <15) for a consistent
#     look across the notebook.
#   - Price is formatted with a leading '$' and two decimal places.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def display_inventory(inventory):
    # print header row with column names
    print(f"{'Product':<30} {'Quantity':<15} {'Price'}")
    print("-" * 55)  # separator line

    # loop through each item and print formatted row
    for product, details in inventory.items():
        print(f"{product:<30} {details['quantity']:<15} ${details['price']:.2f}")

# Confirmation that the cell ran and the function is now available to call.
print("display_inventory function defined and ready to use.")

display_inventory function defined and ready to use.


## Step 4: Calculate Total Inventory Value

Create a function that calculates the total value of all inventory (quantity × price for each item, then sum all items).


In [26]:
# ---------------------------------------------------------------------------
# Function: calculate_total_value(inventory)
#
# Purpose:
#   Calculates the total monetary value of all items currently in the
#   inventory by summing quantity * price for every product.
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#
# Returns:
#   float: the total value of the inventory. Returns 0 if the inventory
#          dictionary is empty.
#
# Notes:
#   - Iterates over every product and accumulates quantity * price.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def calculate_total_value(inventory):
    total_value = 0

    # loop through all items and add quantity x price for each
    for product, details in inventory.items():
        total_value += details['quantity'] * details['price']

    return total_value

# Confirmation that the cell ran and the function is now available to call.
print("calculate_total_value function defined and ready to use.")


calculate_total_value function defined and ready to use.


## Step 5: Find Low Stock Items

Create a function that finds all items with quantity below a specified threshold (e.g., 10 items).


In [27]:
# ---------------------------------------------------------------------------
# Function: find_low_stock_items(inventory, threshold=10)
#
# Purpose:
#   Returns a list of product names whose quantity is below the given
#   threshold, so the user can quickly see which items need restocking.
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#   threshold (int):  quantity cutoff; products with quantity strictly less
#                     than this value are considered low stock. Defaults to 10.
#
# Returns:
#   list[str]: product names that are low in stock. Returns an empty list
#              if no products fall below the threshold (or the inventory
#              is empty).
#
# Notes:
#   - Comparison uses strict less-than (<), so a product with exactly the
#     threshold quantity is NOT included.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def find_low_stock_items(inventory, threshold=10):
    low_stock = []

    # loop through inventory and check each item's quantity
    for product, details in inventory.items():
        if details['quantity'] < threshold:  # below threshold means low stock
            low_stock.append(product)        # add product name to list

    return low_stock

# Confirmation that the cell ran and the function is now available to call.
print("find_low_stock_items function defined and ready to use.")

find_low_stock_items function defined and ready to use.


## Step 6: Add or Update Inventory Item

Create a function that allows adding a new product or updating an existing product's quantity or price.


In [28]:
# ---------------------------------------------------------------------------
# Function: add_or_update_item(inventory, product_name, quantity, price)
#
# Purpose:
#   Adds a new product to the inventory, or updates an existing product's
#   quantity and price if it is already present.
#
# Parameters:
#   inventory (dict):    nested dictionary in the form
#                        {product_name: {'quantity': int, 'price': float}, ...}
#   product_name (str):  name of the product to add or update (used as key).
#   quantity (int):      new quantity value for the product.
#   price (float):       new price value for the product.
#
# Returns:
#   None. The inventory dictionary is modified in place and a confirmation
#   message is printed indicating whether the product was added or updated.
#
# Notes:
#   - If product_name already exists in the inventory, both quantity and
#     price are overwritten with the new values.
#   - If product_name is not in the inventory, a new entry is created.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def add_or_update_item(inventory, product_name, quantity, price):
    if product_name in inventory:
        # product already exists: update quantity and price
        inventory[product_name]['quantity'] = quantity
        inventory[product_name]['price'] = price
        print(f"'{product_name}' has been updated.")
    else:
        # product is new: add it to the inventory
        inventory[product_name] = {'quantity': quantity, 'price': price}
        print(f"'{product_name}' has been added.")

# Confirmation that the cell ran and the function is now available to call.
print("add_or_update_item function defined and ready to use.")


add_or_update_item function defined and ready to use.


## Step 7: Find Most and Least Expensive Items

Create functions to find the most expensive and least expensive items in the inventory.


In [29]:
# ---------------------------------------------------------------------------
# Function: find_most_expensive(inventory)
#
# Purpose:
#   Finds the product with the highest price in the inventory and returns
#   both its name and its price.
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#
# Returns:
#   tuple: (product_name, price) for the most expensive item.
#          Returns (None, 0) if the inventory is empty.
#
# Notes:
#   - Iterates through every product and tracks the highest price seen.
#   - If multiple products share the same highest price, the first one
#     encountered during iteration is returned.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def find_most_expensive(inventory):
    if not inventory:  # return default if inventory is empty
        return (None, 0)

    max_product = None
    max_price = -1  # start low so any price beats it

    for product, details in inventory.items():
        if details['price'] > max_price:  # found a higher price
            max_price = details['price']
            max_product = product

    return (max_product, max_price)


# ---------------------------------------------------------------------------
# Function: find_least_expensive(inventory)
#
# Purpose:
#   Finds the product with the lowest price in the inventory and returns
#   both its name and its price.
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#
# Returns:
#   tuple: (product_name, price) for the least expensive item.
#          Returns (None, 0) if the inventory is empty.
#
# Notes:
#   - Iterates through every product and tracks the lowest price seen,
#     starting from float('inf') so any real price will be smaller.
#   - If multiple products share the same lowest price, the first one
#     encountered during iteration is returned.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def find_least_expensive(inventory):
    if not inventory:  # return default if inventory is empty
        return (None, 0)

    min_product = None
    min_price = float('inf')  # start very high so any price beats it

    for product, details in inventory.items():
        if details['price'] < min_price:  # found a lower price
            min_price = details['price']
            min_product = product

    return (min_product, min_price)


# Confirmation that the cell ran and the functions are now available to call.
print("find_most_expensive function defined and ready to use.")
print("find_least_expensive function defined and ready to use.")


find_most_expensive function defined and ready to use.
find_least_expensive function defined and ready to use.


## Step 8: Generate Inventory Report

Create a function that generates a comprehensive inventory report with statistics.


In [30]:
# ---------------------------------------------------------------------------
# Function: generate_report(inventory)
#
# Purpose:
#   Prints a comprehensive, human-readable inventory report summarizing key
#   statistics about the current inventory (counts, value, extremes, low
#   stock, averages, and totals).
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#
# Returns:
#   None. The report is printed directly to the console.
#
# Notes:
#   - Relies on helper functions defined earlier in the notebook:
#     calculate_total_value, find_most_expensive, find_least_expensive,
#     find_low_stock_items, calculate_average_price, and
#     calculate_total_quantity. Those cells must be run first.
#   - Output is bracketed by '=' separator lines for readability.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def generate_report(inventory):
    print("=" * 55)
    print("INVENTORY REPORT")
    print("=" * 55)

    # total number of products
    print(f"Total number of products:  {len(inventory)}")

    # total inventory value
    total_value = calculate_total_value(inventory)
    print(f"Total inventory value:     ${total_value:.2f}")

    # most expensive item
    most_exp_name, most_exp_price = find_most_expensive(inventory)
    print(f"Most expensive item:       {most_exp_name} (${most_exp_price:.2f})")

    # least expensive item
    least_exp_name, least_exp_price = find_least_expensive(inventory)
    print(f"Least expensive item:      {least_exp_name} (${least_exp_price:.2f})")

    # low stock items
    low_stock = find_low_stock_items(inventory)
    print(f"Low stock items:           {low_stock if low_stock else 'None'}")

    # average price
    avg_price = calculate_average_price(inventory)
    print(f"Average price:             ${avg_price:.2f}")

    # total quantity
    total_qty = calculate_total_quantity(inventory)
    print(f"Total quantity (all items): {total_qty}")

    print("=" * 55)

# Confirmation that the cell ran and the function is now available to call.
print("generate_report function defined and ready to use.")


generate_report function defined and ready to use.


## Step 9: Search for Product

Create a function to search for a product in the inventory by name.


In [31]:
# ---------------------------------------------------------------------------
# Function: search_product(inventory, product_name)
#
# Purpose:
#   Searches the inventory for a product by name and returns its details if
#   found. The search is case-insensitive, so 'apple', 'Apple', and 'APPLE'
#   all match the same product.
#
# Parameters:
#   inventory (dict):    nested dictionary in the form
#                        {product_name: {'quantity': int, 'price': float}, ...}
#   product_name (str):  name of the product to look up.
#
# Returns:
#   dict or None: a single-entry dictionary in the same nested format
#                 ({product_name: {'quantity': int, 'price': float}})
#                 if the product is found; otherwise None.
#
# Notes:
#   - Comparison is done using .lower() on both the stored key and the
#     search term so capitalization differences do not matter.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def search_product(inventory, product_name):
    # case-insensitive search by comparing lowercase versions
    for product, details in inventory.items():
        if product.lower() == product_name.lower():
            # return matching product in same dictionary format
            return {product: {'quantity': details['quantity'], 'price': details['price']}}

    return None  # not found

# Confirmation that the cell ran and the function is now available to call.
print("search_product function defined and ready to use.")


search_product function defined and ready to use.


## Step 10: Remove Item from Inventory

Create a function to remove items from the inventory.


In [46]:
# ---------------------------------------------------------------------------
# Function: remove_item(inventory, product_name)
#
# Purpose:
#   Removes a product from the inventory by name. Prints a confirmation
#   message indicating whether the product was successfully removed or
#   was not found in the inventory.
#
# Parameters:
#   inventory (dict):    nested dictionary in the form
#                        {product_name: {'quantity': int, 'price': float}, ...}
#   product_name (str):  name of the product to remove (used as the key).
#
# Returns:
#   None. The inventory dictionary is modified in place when the product
#   exists; otherwise it is left unchanged.
#
# Notes:
#   - Lookup is case-sensitive (uses 'in inventory'); product_name must
#     match the stored key exactly.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def remove_item(inventory, product_name):
    if product_name in inventory:
        # Capture the removed item's stock BEFORE deleting so we can report
        # how many of that specific product were still on hand.
        removed_quantity = inventory[product_name]['quantity']
        removed_price = inventory[product_name]['price']
        del inventory[product_name]  # remove the product entry
        print(f"'{product_name}' has been removed from inventory.")

        # Show the remaining stock for the item that was just removed.
        # Only print this when the removed item still had units in stock
        # (quantity > 0). Removing the product entry from the dictionary
        # does not actually reduce the physical stock that was on hand.
        if removed_quantity > 0:
            print(f"\nRemaining stock of '{product_name}' still on hand: "
                  f"{removed_quantity} unit(s) @ ${removed_price:.2f} each.")
            print("Note: removing this item only deleted its inventory entry; "
                  f"the {removed_quantity} unit(s) of '{product_name}' that "
                  "were in stock have not been physically removed.")

        # Display the complete remaining product and inventory list as a
        # table to confirm the removed product is no longer present.
        if inventory:
            print("\nComplete remaining inventory after removal (alphabetical order):")
            print(f"{'Product Name':<30} {'Quantity':<15} {'Price'}")
            print("-" * 55)
            # Sort product names alphabetically (case-insensitive) so the
            # remaining inventory table is displayed A -> Z by product name.
            for name in sorted(inventory.keys(), key=str.lower):
                details = inventory[name]
                print(f"{name:<30} {details['quantity']:<15} ${details['price']:.2f}")
        else:
            print("\nThe inventory is now empty.")
    else:
        print(f"'{product_name}' was not found in inventory.")

# Confirmation that the cell ran and the function is now available to call.
print("remove_item function defined and ready to use.")


remove_item function defined and ready to use.


## Step 11: Price Range Filter

Create a function to find all items within a specific price range.


In [33]:
# ---------------------------------------------------------------------------
# Function: find_items_in_price_range(inventory, min_price, max_price)
#
# Purpose:
#   Returns a list of products whose price falls within the specified
#   price range (inclusive on both ends), so the user can filter the
#   inventory by an affordable or target price window.
#
# Parameters:
#   inventory (dict):  nested dictionary in the form
#                      {product_name: {'quantity': int, 'price': float}, ...}
#   min_price (float): minimum price of the range (inclusive).
#   max_price (float): maximum price of the range (inclusive).
#
# Returns:
#   list[tuple]: a list of (product_name, price) tuples for every product
#                whose price is between min_price and max_price. Returns
#                an empty list if no products match (or the inventory is
#                empty).
#
# Notes:
#   - Comparison uses min_price <= price <= max_price, so products priced
#     exactly at the bounds ARE included.
#   - The caller is responsible for passing a sensible range
#     (min_price <= max_price); no validation is performed here.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def find_items_in_price_range(inventory, min_price, max_price):
    results = []  # list to store matching items

    for product, details in inventory.items():
        price = details['price']
        # check if price falls within the specified range (inclusive)
        if min_price <= price <= max_price:
            results.append((product, price))  # add tuple to results

    return results

# Confirmation that the cell ran and the function is now available to call.
print("find_items_in_price_range function defined and ready to use.")


find_items_in_price_range function defined and ready to use.


## Step 12: Calculate Statistics

Create functions to calculate statistics about the inventory.


In [34]:
# ---------------------------------------------------------------------------
# Function: calculate_average_price(inventory)
#
# Purpose:
#   Calculates the average (mean) price across all products in the
#   inventory so the user can see a quick summary statistic of pricing.
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#
# Returns:
#   float: the average price of all items in the inventory.
#          Returns 0.0 if the inventory is empty (avoids divide-by-zero).
#
# Notes:
#   - Average is computed as (sum of all prices) / (number of products);
#     quantities are NOT weighted into this average.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def calculate_average_price(inventory):
    """
    Calculate the average price of all items in inventory.

    Parameters:
    inventory (dict): The inventory dictionary

    Returns:
    float: The average price, or 0.0 if inventory is empty
    """
    # return 0.0 for empty inventory to avoid divide-by-zero
    if len(inventory) == 0:
        return 0.0

    total = 0  # running sum of all prices
    for product, details in inventory.items():
        total += details['price']  # add each product's price

    return total / len(inventory)  # mean price

# ---------------------------------------------------------------------------
# Function: calculate_total_quantity(inventory)
#
# Purpose:
#   Calculates the total number of units across all products in the
#   inventory by summing each product's quantity.
#
# Parameters:
#   inventory (dict): nested dictionary in the form
#                     {product_name: {'quantity': int, 'price': float}, ...}
#
# Returns:
#   int: the combined quantity of every product in the inventory.
#        Returns 0 if the inventory is empty.
#
# Notes:
#   - Only the 'quantity' field is summed; prices are ignored here.
#   - Defining this cell only creates the function; no output is produced
#     until the function is actually called.
# ---------------------------------------------------------------------------
def calculate_total_quantity(inventory):
    """
    Calculate the total quantity of all items in inventory.

    Parameters:
    inventory (dict): The inventory dictionary

    Returns:
    int: The total quantity of all items
    """
    total = 0  # running sum of all quantities
    for product, details in inventory.items():
        total += details['quantity']  # add each product's quantity

    return total

# Confirmation that the cell ran and the functions are now available to call.
print("calculate_average_price function defined and ready to use.")
print("calculate_total_quantity function defined and ready to use.")


calculate_average_price function defined and ready to use.
calculate_total_quantity function defined and ready to use.


## Step 13: Interactive Menu System

This interactive menu system using a while loop that allows users to:
- View inventory
- Add items
- Update items
- Remove items
- Search for products
- View reports
- Exit and Clear Past Entries


In [35]:
from IPython.display import display, clear_output, HTML
import ipywidgets as widgets

# Inject CSS so the interactive menu widgets (dropdown, text inputs, buttons,
# labels) use the same monospace font as the notebook's code/output cells.
# This keeps the visual style consistent between the menu and the printed
# tables/reports that appear below it.
display(HTML("""
<style>
.widget-label, .widget-dropdown, .widget-dropdown select,
.widget-text input, .widget-inttext input, .widget-floattext input,
.widget-button, .jupyter-button {
    font-family: monospace, "Courier New", Courier !important;
    font-size: 13px !important;
}
</style>
"""))

def run_interactive_menu(inventory):
    """
    Run an interactive menu system using IPython widgets.
    This avoids the clear_output/input prompt issues in Jupyter.
    """
    output = widgets.Output()

    def handle_menu_choice(choice):
        with output:
            clear_output(wait=True)


            if choice == '1':
                print("\n" + "="*60)
                print("LOAD INVENTORY")
                print("="*60)
                # Reload the inventory from the CSV file in place so that
                # the same dictionary object the menu was given keeps
                # working for all other options.
                reloaded = load_inventory(CSV_FILENAME)
                inventory.clear()
                inventory.update(reloaded)
                print("="*60)
                print(f"Inventory reloaded: {len(inventory)} products.")
                print("="*60)

            elif choice == '2':
                print("\n" + "="*60)
                print("SEARCH FOR PRODUCT")
                print("="*60)

                name_input = widgets.Text(description='Product Name:', style={'description_width': 'initial'}, layout=widgets.Layout(width='400px'))
                search_btn = widgets.Button(description='Search', button_style='info')
                cancel_btn = widgets.Button(description='Cancel', button_style='')

                def on_search(b):
                    with output:
                        clear_output(wait=True)

                        result = search_product(inventory, name_input.value)
                        if result:
                            print("\n" + "=" * 60)
                            print("PRODUCT FOUND")
                            print("=" * 60)
                            display_inventory(result)
                            print("=" * 60)
                        else:
                            print("\n" + "=" * 60)
                            print(f"PRODUCT '{name_input.value}' NOT FOUND.")
                            print("=" * 60)

                def on_cancel(b):
                    with output:
                        clear_output(wait=True)


                search_btn.on_click(on_search)
                cancel_btn.on_click(on_cancel)

                display(widgets.VBox([name_input, widgets.HBox([search_btn, cancel_btn])]))
                return

            elif choice == '3':
                print("\n" + "="*60)
                print("COMPLETE INVENTORY")
                print("="*60)
                display_inventory(inventory)
                print("="*60)

            elif choice == '4':
                print("\n" + "="*60)
                print("ADD/UPDATE ITEM")
                print("="*60)

                # Create input widgets
                name_input = widgets.Text(description='Product Name:', style={'description_width': 'initial'}, layout=widgets.Layout(width='400px'))
                qty_input = widgets.IntText(description='Quantity:', style={'description_width': 'initial'}, layout=widgets.Layout(width='400px'))
                price_input = widgets.FloatText(description='Price:', style={'description_width': 'initial'}, layout=widgets.Layout(width='400px'))
                submit_btn = widgets.Button(description='Submit', button_style='success')
                cancel_btn = widgets.Button(description='Cancel', button_style='')

                def on_submit(b):
                    with output:
                        clear_output(wait=True)
                        print("\n" + "="*60)
                        print("ITEM ADDED/UPDATED SUCCESSFULLY")
                        print("="*60)
                        add_or_update_item(inventory, name_input.value, qty_input.value, price_input.value)
                        print("="*60)

                def on_cancel(b):
                    with output:
                        clear_output(wait=True)


                submit_btn.on_click(on_submit)
                cancel_btn.on_click(on_cancel)

                display(widgets.VBox([name_input, qty_input, price_input,
                                      widgets.HBox([submit_btn, cancel_btn])]))
                return

            elif choice == '5':
                print("\n" + "="*60)
                print("REMOVE ITEM")
                print("="*60)

                name_input = widgets.Text(description='Product Name:', style={'description_width': 'initial'}, layout=widgets.Layout(width='400px'))
                submit_btn = widgets.Button(description='Remove', button_style='danger')
                cancel_btn = widgets.Button(description='Cancel', button_style='')

                def on_submit(b):
                    with output:
                        clear_output(wait=True)
                        print("\n" + "="*60)
                        remove_item(inventory, name_input.value)
                        print("="*60)

                def on_cancel(b):
                    with output:
                        clear_output(wait=True)


                submit_btn.on_click(on_submit)
                cancel_btn.on_click(on_cancel)

                display(widgets.VBox([name_input, widgets.HBox([submit_btn, cancel_btn])]))
                return

            elif choice == '6':
                print("\n" + "="*60)
                print("FIND ITEMS IN PRICE RANGE")
                print("="*60)

                min_input = widgets.FloatText(description='Min Price:', style={'description_width': 'initial'}, layout=widgets.Layout(width='400px'))
                max_input = widgets.FloatText(description='Max Price:', style={'description_width': 'initial'}, layout=widgets.Layout(width='400px'))
                search_btn = widgets.Button(description='Search', button_style='info')
                cancel_btn = widgets.Button(description='Cancel', button_style='')

                def on_search(b):
                    with output:
                        clear_output(wait=True)

                        results = find_items_in_price_range(inventory, min_input.value, max_input.value)
                        if results:
                            print("\n" + "=" * 60)
                            print("FOUND ITEMS")
                            print("=" * 60)
                            print(f"\nItems in price range ${min_input.value:.2f} - ${max_input.value:.2f}:")
                            for product_name, price in results:
                                print(f"  {product_name}: ${price:.2f}")
                            print("="*60)
                        else:
                            print("\n" + "=" * 60)
                            print("NO ITEMS FOUND IN THAT PRICE RANGE")
                            print("=" * 60)

                def on_cancel(b):
                    with output:
                        clear_output(wait=True)


                search_btn.on_click(on_search)
                cancel_btn.on_click(on_cancel)

                display(widgets.VBox([min_input, max_input, widgets.HBox([search_btn, cancel_btn])]))
                return

            elif choice == '7':
                print("\n" + "=" * 60)
                print("INVENTORY REPORT")
                print("=" * 60)
                generate_report(inventory)
                print("=" * 60)



    # Create menu dropdown
    menu_dropdown = widgets.Dropdown(
        options=[
            ('Select an option...', ''),
            ('1. Load Inventory', '1'),
            ('2. Search for Product', '2'),
            ('3. View Inventory', '3'),
            ('4. Add/Update Item', '4'),
            ('5. Remove Item', '5'),
            ('6. Find Items in Price Range', '6'),
            ('7. View Report', '7')

        ],
        description='Menu:',
        value='',
        style={'description_width': 'initial'},
        layout=widgets.Layout(width='400px')
    )

    # Handle menu selection
    def on_menu_change(change):
        if change['new']:  # Only process if a selection was made
            handle_menu_choice(change['new'])
            menu_dropdown.value = ''  # Reset dropdown

    menu_dropdown.observe(on_menu_change, names='value')

    # Display the menu and output area
    display(widgets.VBox([menu_dropdown, output]))

# Confirmation that the cell ran and the function is now available to call.
print("run_interactive_menu function defined and ready to use.")

run_interactive_menu function defined and ready to use.


## Step 14: Main Program

1. Loads the inventory from the CSV file
2. Runs the interactive menu system
3. Saves the updated inventory back to the file when the user exits


In [47]:
## Main program
def main():
    """
    Main function to run the inventory management system.
    """
    # Use the CSV_FILENAME variable defined at the top of the notebook
    filename = CSV_FILENAME

    # Load inventory from file
    print("Loading inventory...")
    inventory = load_inventory(filename)

    if inventory:
        print(f"Successfully loaded {len(inventory)} products.")
        # Launch the interactive dropdown menu so the user can view, add,
        # update, remove, search, report on, and filter inventory items.
        run_interactive_menu(inventory)

# Confirmation that the cell ran and the function is now available to call.
print("main function defined and ready to use.")

# Actually call main() so the program runs when this cell is executed.
main()

main function defined and ready to use.
Loading inventory...
File 'Data/inventory_data.csv' opened successfully.
Product Name              Processing                Quantity   Price      Status
-----------------------------------------------------------------------------------------------
Laptop                    Processing product        15         $899.99    Added to inventory
Mouse                     Processing product        45         $24.99     Added to inventory
Keyboard                  Processing product        32         $49.99     Added to inventory
Monitor                   Processing product        8          $299.99    Added to inventory
Headphones                Processing product        12         $79.99     Added to inventory
Webcam                    Processing product        5          $89.99     Added to inventory
USB Drive                 Processing product        50         $12.99     Added to inventory
External Hard Drive       Processing product        18     

Midterm Project Questions:

**Question 1:** Why is it important to use try-except blocks when reading from and writing to CSV files? What specific exceptions should you handle, and what could go wrong if you don't handle them?

**Answer:** Using try-except blocks is really important when working with CSV files because a lot of things can go wrong that you cant always predict. The biggest one is FileNotFoundError, which happens when the file path is wrong or the file just isnt there yet. If you dont catch that, the whole program just crashes with a weird error message that doesnt help the user at all. Other things that can go wrong include not having permission to open the file, or the data inside being formatted wrong so it cant be converted properly. By wrapping the file reading code in a try-except, the program can handle those situations gracefully and let the user know what happened instead of just stopping.

**Question 2:** What are the advantages of using Python's csv module to read CSV files instead of reading the file as plain text and splitting by commas? Provide a specific example where the csv module would handle data correctly but plain text reading would fail.

**Answer:** The csv module is way more reliable than just doing line.split(',') on every line. The biggest problem with plain text splitting is that it breaks as soon as a value contains a comma. For example if you had a product called "Hard Drive, External" and you tried to split that line by comma, you would end up with too many pieces and your indexes would be completely off. The csv module knows to look for quoted fields and treats everything inside the quotes as one single value, so it handles that situation perfectly. It also deals with things like newlines inside fields and different types of delimiters, which plain text reading just cant do.

**Question 3:** When reading data from a CSV file, all values are initially strings. Explain why it's necessary to convert the quantity to an int and the price to a float. What would happen if you tried to perform mathematical operations without converting these values first?

**Answer:** When Python reads a CSV file every single value comes in as a string, even if it looks like a number. So something like 15 is actually the string "15" and 899.99 is the string "899.99". If you try to multiply those together to get a total value, Python will throw a TypeError because you cant do math on strings. Even something like adding two quantities together would give you the wrong answer since "15" + "5" in Python gives you "155" instead of 20 because it just concatenates the strings. Converting quantity to int and price to float right when you read the data makes sure all the math in the rest of the program works the way it should.

**Question 4:** Explain the benefits of breaking the inventory management system into separate functions rather than writing everything in a single main() function. How does this approach make the code easier to test, debug, and maintain?

**Answer:** Breaking everything into separate functions makes the project a lot easier to work with. When I was building this, having separate functions like load_inventory and calculate_total_value meant that if something wasnt working I could test just that one function without having to run the whole program. It also makes the code much easier to read because each function has one clear job and you can tell what it does just from its name. Another big advantage is that you can reuse functions in multiple places. For example calculate_total_value gets called inside generate_report but could also be called anywhere else without having to rewrite that logic. If I had put everything in one big main function it would have been really hard to find bugs or make changes without breaking something else.

**Question 5:** Explain why a dictionary is the most appropriate data structure for storing the inventory data, compared to using a list or tuple. What advantages does a dictionary provide for this specific use case?

**Answer:** A dictionary made the most sense for storing the inventory because the whole point is to look things up by product name. With a dictionary you can just write inventory["Laptop"] and instantly get all the information about that product without having to loop through anything. If I had used a list I would have had to search through every single item every time I wanted to find one product, which gets slow as the inventory grows. A tuple wasnt even really an option because tuples are immutable, meaning you cant change the values once they are set, and an inventory system obviously needs to be able to update quantities and prices. The nested dictionary format where each product name maps to another dictionary with quantity and price also keeps the data really organized and easy to understand.